File: examples/cpp/cancel_request.cpp
/**
 * @file cancel_request.cpp
 * @brief Demonstrates cancelling an in-flight I/O operation (C++)
 *
 * Shows how to use engine.cancel() to abort a pending read.
 * The cancelled operation's callback receives result = -ECANCELED.
 *
 * Usage: ./cancel_request <file>
 */

#include <auraio.hpp>

#include <iostream>
#include <atomic>
#include <fcntl.h>
#include <unistd.h>
#include <cstring>
#include <cerrno>

constexpr size_t READ_SIZE = 4096;

int main(int argc, char *argv[]) {
    if (argc < 2) {
        std::cerr << "Usage: " << argv[0] << " <file>\n";
        return 1;
    }

    try {
        // Create engine with RAII
        auraio::Engine engine;

        // Open file
        int fd = open(argv[1], O_RDONLY);
        if (fd < 0) {
            throw std::runtime_error(std::string("Failed to open file: ") + strerror(errno));
        }

        // Allocate buffer (automatically managed)
        auto buf = engine.allocate_buffer(READ_SIZE);

        // Completion state with atomic for thread-safety
        std::atomic<bool> completed{false};
        std::atomic<ssize_t> read_result{0};

        // Submit async read with lambda callback
        std::cout << "Submitting async read...\n";
        auto req = engine.read(
            fd, buf, READ_SIZE, 0, [&completed, &read_result](auraio::Request &, ssize_t result) {
                read_result.store(result, std::memory_order_release);
                completed.store(true, std::memory_order_release);

                if (result == -ECANCELED) {
                    std::cout << "Read was cancelled (result = -ECANCELED)\n";
                } else if (result < 0) {
                    std::cerr << "Read failed: " << strerror(static_cast<int>(-result)) << "\n";
                } else {
                    std::cout << "Read completed: " << result << " bytes\n";
                }
            });

        /* Attempt to cancel the request.
         * Note: cancellation is best-effort. If the I/O already completed
         * by the time cancel is processed, the original result is returned
         * instead of -ECANCELED. */
        std::cout << "Attempting to cancel...\n";
        bool cancelled = engine.cancel(req);
        if (cancelled) {
            std::cout << "Cancel request submitted successfully\n";
        } else {
            std::cout << "Cancel submission failed (operation may have already completed)\n";
        }

        // Wait for completion callback
        while (!completed.load(std::memory_order_acquire)) {
            engine.wait(100); // 100ms timeout
        }

        ssize_t final_result = read_result.load(std::memory_order_acquire);
        std::cout << "Final result: " << final_result << " ("
                  << (final_result == -ECANCELED ? "cancelled"
                      : final_result < 0         ? strerror(static_cast<int>(-final_result))
                                                 : "success")
                  << ")\n";

        // Cleanup (automatic via RAII)
        close(fd);
        return 0;

    } catch (const std::exception &e) {
        std::cerr << "Error: " << e.what() << "\n";
        return 1;
    }
}

File: examples/rust/examples/cancel_request.rs
// SPDX-License-Identifier: Apache-2.0
// Copyright 2026 AuraIO Contributors


//! Demonstrates cancelling an in-flight I/O operation
//!
//! Shows how to use engine.cancel() to abort a pending read.
//! The cancelled operation's callback receives Err(Error::Cancelled).
//!
//! Usage: cargo run --example cancel_request -- <file>

use aura::{Engine, Result};
use std::env;
use std::fs::File;
use std::os::unix::io::AsRawFd;
use std::sync::atomic::{AtomicBool, AtomicIsize, Ordering};
use std::sync::Arc;

const READ_SIZE: usize = 4096;

fn main() -> Result<()> {
    let args: Vec<String> = env::args().collect();
    if args.len() < 2 {
        eprintln!("Usage: {} <file>", args[0]);
        std::process::exit(1);
    }

    // Create engine
    let engine = Engine::new()?;

    // Open file
    let file = File::open(&args[1])?;
    let fd = file.as_raw_fd();

    // Allocate buffer
    let buf = engine.allocate_buffer(READ_SIZE)?;

    // Completion state with Arc for safe sharing
    let completed = Arc::new(AtomicBool::new(false));
    let read_result = Arc::new(AtomicIsize::new(0));

    let completed_clone = Arc::clone(&completed);
    let result_clone = Arc::clone(&read_result);

    // Submit async read with closure callback
    println!("Submitting async read...");
    let req = unsafe {
        engine.read(
            fd,
            (&buf).into(),
            READ_SIZE,
            0,
            move |result| {
                match result {
                    Ok(n) => {
                        println!("Read completed: {} bytes", n);
                        result_clone.store(n as isize, Ordering::SeqCst);
                    }
                    Err(e) => {
                        match e {
                            aura::Error::Cancelled => {
                                println!("Read was cancelled (result = -ECANCELED)");
                                result_clone.store(-(libc::ECANCELED as isize), Ordering::SeqCst);
                            }
                            aura::Error::Io(io_err) => {
                                eprintln!("Read failed: {}", io_err);
                                let err_code = io_err.raw_os_error().unwrap_or(-1);
                                result_clone.store(-(err_code as isize), Ordering::SeqCst);
                            }
                            _ => {
                                eprintln!("Read failed: {}", e);
                                result_clone.store(-1, Ordering::SeqCst);
                            }
                        }
                    }
                }
                completed_clone.store(true, Ordering::SeqCst);
            },
        )?
    };

    // Attempt to cancel the request.
    // Note: cancellation is best-effort. If the I/O already completed
    // by the time cancel is processed, the original result is returned
    // instead of Err(Error::Cancelled).
    println!("Attempting to cancel...");
    unsafe {
        match engine.cancel(&req) {
            Ok(_) => println!("Cancel request submitted successfully"),
            Err(_) => println!("Cancel submission failed (operation may have already completed)"),
        }
    }

    // Wait for completion callback
    while !completed.load(Ordering::SeqCst) {
        engine.wait(100)?; // 100ms timeout
    }

    let final_result = read_result.load(Ordering::SeqCst);
    let status_str = if final_result == -(libc::ECANCELED as isize) {
        "cancelled"
    } else if final_result < 0 {
        "error"
    } else {
        "success"
    };

    println!("Final result: {} ({})", final_result, status_str);

    Ok(())
}
